README.md
# Query Parallelizer

## What it is

`qp` is a performance-boosting proxy for sharded MySQL databases. It recieves requests containing multiple queries, makes those queries in parallel, and returns the aggregated results. Requests are JSON formatted as an array of objects, each with a destination [DSN](http://en.wikipedia.org/wiki/Data_source_name) and query. Queries run in parallel for each request and requests are handled in parallel, for maximum concurrency.

An example request and response:

**Request**
```
[
  {
    "dsn": "root@tcp(127.0.0.1:20890)/product?charset=utf8",
    "query": "select pid, views from product_view order by rand() limit 2;"
  },
  {
    "dsn": "root@tcp(127.0.0.1:20891)/product?charset=utf8",
    "query": "select pid, views from product_view order by rand() limit 2;"
  },
  {
    "dsn": "root@tcp(127.0.0.1:20892)/product?charset=utf8",
    "query": "select pid, views from product_view order by rand() limit 2;"
  },
  {
    "dsn": "root@tcp(127.0.0.1:20893)/product?charset=utf8",
    "query": "select pid, views from product_view order by rand() limit 2;"
  }
]
```

**Response**
```
{
  "root@tcp(127.0.0.1:20890)/product?charset=utf8": [
    ["3732", 471], ["21017", 1737]
  ],
  "root@tcp(127.0.0.1:20891)/product?charset=utf8": [
    ["4111", 23098], ["16701", 23098]
  ],
  "root@tcp(127.0.0.1:20892)/product?charset=utf8": [
    ["2474", 273], ["22118", 1211]
  ],
  "root@tcp(127.0.0.1:20893)/product?charset=utf8": [
    ["2920", 415], ["25870", 1980]
  ]
}
```

**Note:** This response format implies that if a request contains multiple queries for the same DSNs, only one of the result sets will be returned per DSN, specifically the last one that is returned by that DSN. There may be a "tag" parameter in the future that would allow a request to contain multiple queries for the same database.

## Why it's different

Other shard-aggregating MySQL proxies include [Shard-Query](https://github.com/greenlion/swanhart-tools/tree/master/shard-query) and [Spock Proxy](http://spockproxy.sourceforge.net/). Both of these projects require an auxiliary MySQL database to be created and populated with shard configuration information. This is neccessary because these packages take a query with cross-shard predicates, and re-write the query internally for each shard. 

**This has a number of disadvantages:**

* The database-backed configuration is a hassle to create, operate, and maintain, and imposes limits on the sharded infrastructure that `qp` does not.
* Shard-key logic is limited to the options available in these packages. Removing this limitation was one of the main drivers behind creating `qp`: a database was sharded using a digest of multiple column values, and existing packages did not support this technique.
* The overhead of re-executing query-planning logic in these systems adds considerable overhead to each request, so much so that Shard-Query indicates that it is not meant for OLTP, with an expected minimum query time of 20ms.

**`qp` offers an alternative:**

* It is configuration free. Connection pools are opened to the first N DSNs that are requested, enabling one `qp` process to support many arbitrary pool configurations.
* Shard-key logic is offloaded to the client, which may even use `qp` to aggregate data from different schemas.
* `qp` does one thing and does it well, parallelize queries. Queries are not re-written so there is minimal service overhead.
* `qp` is meant to be used online for high-request-throughput workloads.

## Why it's fast

On startup, `qp` spawns a number of concurrent workers, each of which can make concurrent queries per request. Concurrent workers and queries are goroutines, not threads, which means that there is **no OS setup cost or context switching overhead in the concurrency model**. In fact, I/O syscalls made by goroutines are automatically multiplexed by Golang using the best available method on the system, e.g. [epoll on Linux or kqueue on BSD/OSX](https://groups.google.com/d/msg/golang-nuts/AQ8JOHxm9jA/cakNJj7_BVkJ). This means that Golang determines the appropraite number of "real" OS threads to use for parallel execution, while scheduling a collection of [coroutines](http://en.wikipedia.org/wiki/Coroutine) that it calls goroutines. To take advantage of performance enhancements to the Golang scheulder, **the latest version of Golang should be used. `qp` has been tested with Go version 1.4.2.**

As for MySQL connections, each time a request is made to a new DSN, a connection pool is opened to that server that is shared among all goroutines. Each connection pool has a limit on its number of connections.

## To Build

`make deps && make`

## To Test

`make test`

## To Run

`./qp `

**Options (with default argument shown, if any):**

**`-version`** print the version of `qp` and exit  
**`-url`** `tcp://127.0.0.1:9666` nanomsg URL to use for listening socket  
**`-workers`** `6` number of query workers  
**`-max_dsns`** `16` maximum number of DSNs for which to open a connection pool, once this limit is reached, queries cannot be made to any other DSN  

Example usage with all runtime options:

`./qp -url ipc:///tmp/qp.ipc -workers 64 -max_dsns 256`



ChangeLog
2016-04-18 version 1.0.0
	* Initial release


tests/test.sh
#!/bin/bash

USER_AND_HOST=root@$DEVBOX

function run_test {
    echo "$1"
    $1 2>&1 1>/dev/null
    if [ $? ]; then
        echo -e "PASS\n"
    else
        echo -e "FAIL\n"
    fi
}

run_test "./tests/integration_test.sh $USER_AND_HOST" quiet
run_test "./tests/load_test.sh $USER_AND_HOST" quiet


tests/load_test.sh
#!/bin/bash

GOPATH=`pwd`/.go

PROCESS=./qp
CONCURRENT=10
MAX_DSNS=8
PORT=9666
URL=:$PORT

function cleanup {
    kill $PID
    kill $SSH_TUNNEL_PID
}
trap cleanup EXIT

function die {
    echo $1
    exit 666
}

# If called with no arguments a new timer is returned.
# If called with arguments the first is used as a timer
# value and the elapsed time is returned in the form HH:MM:SS.
#
function timer()
{
    if [[ $# -eq 0 ]]; then
        echo $(date '+%s')
    else
        local  stime=$1
        etime=$(date '+%s')

        if [[ -z "$stime" ]]; then stime=$etime; fi

        dt=$((etime - stime))
        ds=$((dt % 60))
        dm=$(((dt / 60) % 60))
        dh=$((dt / 3600))
        printf '%d:%02d:%02d' $dh $dm $ds
    fi
}

QUERY="$(cat tests/load_test_query.sql)"

TUNNEL_HOST=$1
if [[ ! $TUNNEL_HOST ]]; then
    die "Please supply username@hostname for MySQL tunnels"
fi
TUNNELS=""
QS=()
START_PORT=20889
for i in $(seq 1 $MAX_DSNS); do
    port=$((START_PORT + i))
    Q="{\"dsn\":\"root@tcp(127.0.0.1:$port)/monetate_session?charset=utf8\",\"query\":\"__QUERY__\"}"
    QS=(${QS[@]} $Q)    
    Q="{\"dsn\":\"root@tcp(127.0.0.1:$port)/monetate_session?charset=utf8\",\"query\":\"__QUERY__\"}"
    QS=(${QS[@]} $Q)
    TUNNELS="$TUNNELS -L$port:127.0.0.1:3306"
done

echo "Opening ssh tunnels to mysql..."
ssh -N $TUNNELS $TUNNEL_HOST &
SSH_TUNNEL_PID=$!
echo "Started ssh tunnel to $TUNNEL_HOST with pid:$SSH_TUNNEL_PID"

ALL_Q=${QS[@]}
SHARDS=$(echo "${ALL_Q// /|}" | sed -e "s/__QUERY__/$QUERY/g")

QUIET=$2
if [ QUIET ]; then
    $PROCESS -url $URL -maxDsns=$MAX_DSNS -maxConnsPerDsn=24 2>&1 1>/dev/null &
else
    $PROCESS -url $URL -maxDsns=$MAX_DSNS -maxConnsPerDsn=24 2>&1 1>/dev/null &
fi
PID=$!
echo "Started qp process pid:$PID"
sleep 1

ITERATIONS=5

tmr=$(timer)
SHARDQ=$(echo $SHARDS | tr "|" ",")
for i in $(seq 1 $ITERATIONS); do
    for j in $(seq 1 $CONCURRENT); do
        Q="{\"flat\":true,\"queries\":[$SHARDQ]}"
        python -c "import requests; requests.post('http://localhost:$PORT/', '$Q')" &
        PIDS[${j}]=$!
    done
    echo "iteration process ids: ${PIDS[*]}"
    for p in ${PIDS[*]}; do
        wait $p
    done
    unset PIDS
done
T1=$(timer $tmr)

tmr=$(timer)
for i in $(seq 1 $ITERATIONS); do
    OLDIFS=$IFS
    IFS=$'|'
    for j in $SHARDS; do
        for k in $(seq -s "|" 1 $CONCURRENT); do
            Q="{\"flat\":true,\"queries\":[$j]}"
            python -c "import requests; requests.post('http://localhost:$PORT/', '$Q')" &
            PIDS[${k}]=$!
        done
        echo "iteration process ids: ${PIDS[*]}"
        for p in ${PIDS[*]}; do
            wait $p
        done
        unset PIDS
    done
    IFS=$OLDIFS
done
T2=$(timer $tmr)

if [[ "$T1" < "$T2" ]]; then
    echo "Passed: parallel time: $T1, sequential time: $T2"
else
    die "Failed: parallel time: $T1, sequential time: $T2"
fi


tests/integration_test.sh
#!/bin/bash

# Test for correct behavior.
# This does not test performance.


GOPATH=`pwd`/.go

PORT=9666
URL=:$PORT

TEST_DB="qp_test_database"
PID=-1
SSH_TUNNEL_PID=-1

function die()
{
    echo $1
    shutdown
    echo "FAIL"
    exit 666
}

MYSQL_HOST=$1
if [ "$MYSQL_HOST" == "" ]; then
    echo "Please supply username@hostname for MySQL tunnels"
    exit 666
fi


function create_test_database()
{
    mysql --host=127.0.0.1 --user=root --port=22890 -e "drop database if exists qp_test_database;"
    mysqladmin --host=127.0.0.1 --port=22890 -u root create $TEST_DB
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "drop table if exists test;"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "create table test(id int, type varchar(64), value varchar(64));"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(1, 'A', '123');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(2, 'B', '123');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(3, 'C', '123');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(4, 'D', '456');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(5, 'E', '456');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(6, 'F', '456');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(7, 'G', '456');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(8, 'H', '567');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(9, 'I', '567');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(10, 'J', '567');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(11, 'K', '567');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(12, 'L', '678');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(13, 'M', '678');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(14, 'N', '678');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(15, 'O', '678');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(16, 'P', '789');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(17, 'Q', '789');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(18, 'R', '789');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(19, 'S', '789');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(20, 'T', '890');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(21, 'U', '890');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(22, 'V', '890');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(23, 'W', '890');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(24, 'X', '234');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(25, 'Y', '234');"
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(26, 'Z', '234');"
    
    mysql --host=127.0.0.1 --user=root --port=22890 $TEST_DB -e "insert into test values(100, NULL, '999');"
    echo "created test database"
}

function create_local_tunnels()
{
    ip=127.0.0.1
    ssh -Nn -L 22890:$ip:3306 -L 22891:$ip:3306 -L 22892:$ip:3306 -L 22893:$ip:3306 $MYSQL_HOST &
    SSH_TUNNEL_PID=$!
    # wait for tunnel to be established
    while $(netstat -anp tcp | awk '$6 == "LISTEN" && $4 ~ "\.22890"') > /dev/null 2>&1; do
        echo "Waiting for tunnel to be up..."
        sleep 0.5
    done
    echo "Tunnel is up."
}

QUET=$2
function create_qp()
{
    PROCESS=./qp
    MAX_DSNS=8
    if [ QUIET ]; then
        $PROCESS -url $URL -maxDsns=$MAX_DSNS -maxConnsPerDsn=24 2>&1 1>/dev/null &
    else
        $PROCESS -url $URL -maxDsns=$MAX_DSNS -maxConnsPerDsn=24 &
    fi
    PID=$!
    sleep 2
}

function startup()
{
    echo "starting up"
    create_local_tunnels
    create_test_database
    create_qp
}

function shutdown()
{
    echo "shutting down"
    kill $SSH_TUNNEL_PID
    kill $PID
}

function single_query_test()
{
    Q="{\"flat\":true,\"queries\":[{\"dsn\":\"root@tcp(127.0.0.1:22890)/$TEST_DB?charset=utf8\",\"query\":\"select type from test where value=\\\\\"234\\\\\" order by type\"}]}"
    EXPECTED='[["type"],["NullString"],["X"],["Y"],["Z"]]'
    RESPONSE=$(python -c "import requests; print requests.post('http://localhost:$PORT/', '$Q').text")
    if [ "$RESPONSE" != "$EXPECTED" ]; then
        die "single_query_test bad response: $RESPONSE, expected $EXPECTED"
    fi
    echo "single_query_test: pass"
}

function multi_query_test()
{
    Q1="{\"dsn\":\"root@tcp(127.0.0.1:22890)/$TEST_DB?charset=utf8\",\"query\":\"select type from test where value=\\\\\"234\\\\\"\"}"
    Q2="{\"dsn\":\"root@tcp(127.0.0.1:22891)/$TEST_DB?charset=utf8\",\"query\":\"select type from test where value=\\\\\"234\\\\\"\"}"
    Q="{\"flat\":true,\"queries\":[$Q1,$Q2]}"
    EXPECTED='[["type"],["NullString"],["X"],["Y"],["Z"],["X"],["Y"],["Z"]]'
    RESPONSE=$(python -c "import requests; print requests.post('http://localhost:$PORT/', '$Q').text")
    if [ "$RESPONSE" != "$EXPECTED" ]; then
        die "multi_query_test bad response: $RESPONSE, expected $EXPECTED"
    fi
    echo "multi_query_test: pass"
}

function multi_query_mixed_test()
{
    Q1="{\"dsn\":\"root@tcp(127.0.0.1:22890)/$TEST_DB?charset=utf8\",\"query\":\"select type from test where value=\\\\\"234\\\\\"\"}"
    Q2="{\"dsn\":\"root@tcp(127.0.0.1:22891)/$TEST_DB?charset=utf8\",\"query\":\"select type from test where value=\\\\\"123\\\\\"\"}"
    Q="{\"flat\":false,\"queries\":[$Q1,$Q2]}"
    EXPECTED1='[["X"],["Y"],["Z"]]'
    EXPECTED2='[["A"],["B"],["C"]]'
    RESPONSE=$(python -c "import requests; print requests.post('http://localhost:$PORT/', '$Q').text")
    DATA1=$(echo $RESPONSE | jq -a -c '."root@tcp(127.0.0.1:22890)/'$TEST_DB'?charset=utf8"')
    DATA2=$(echo $RESPONSE | jq -a -c '."root@tcp(127.0.0.1:22891)/'$TEST_DB'?charset=utf8"')    
    if [ "$DATA1" != "$EXPECTED1" ]; then
        die "multi_query_mixed_test bad response: $DATA1, expected $EXPECTED1"
    fi
    if [ "$DATA2" != "$EXPECTED2" ]; then
        die "multi_query_mixed_test bad response: $DATA2, expected $EXPECTED2"
    fi
    echo "multi_query_mixed_test: pass"
}

function null_value_test()
{
    Q="{\"flat\":false,\"queries\":[{\"dsn\":\"root@tcp(127.0.0.1:22890)/$TEST_DB?charset=utf8\",\"query\":\"select type from test where value=\\\\\"999\\\\\"\"}]}"
    EXPECTED='[[null]]'
    RESPONSE=$(python -c "import requests; print requests.post('http://localhost:$PORT/', '$Q').text")
    DATA=$(echo $RESPONSE | jq -a -c '."root@tcp(127.0.0.1:22890)/'$TEST_DB'?charset=utf8"')    
    if [ "$DATA" != "$EXPECTED" ]; then
        die "null_value_test bad response: $DATA, expected $EXPECTED"
    fi
    echo "null_value_test: pass"
}

function run_tests()
{    
    single_query_test
    null_value_test
    multi_query_test
    multi_query_mixed_test
}

function handle_interrupt()
{
    die "interrupted"
}

trap 'handle_interrupt' HUP INT QUIT TERM ERR

startup
run_tests
shutdown
echo "SUCCESS"


utils/testquery_sqlalchemy.py
#!/usr/bin/env python

import json
import requests
import sqlalchemy
from decimal import Decimal

query = "select 0.12345678901234567890;"
print "query:", query
print ''

print "# using qp"
qp_query = ("""{{
  "flat": true,
  "queries": [
    {{
      "dsn": "root@tcp(127.0.0.1:3306)/test",
      "query": "{query}"
    }}
  ]
}}""").format(query=query)
response = requests.post("http://127.0.0.1:9666", qp_query)
response = json.loads(response.text)
field_names = response[0]
field_types = response[1]
results = response[2:]
if "DecimalString" in field_types:
    indices = [i for i, x in enumerate(field_types) if x == "DecimalString"]
    for result in results:
        for i in indices:
            result[i] = Decimal(result[i])
print 'parsed response:', results
print ''

print "# using sqlalchemy"
engine = sqlalchemy.create_engine("mysql://root@localhost/test")
conn = engine.connect()
results = conn.execute(sqlalchemy.text(query))
print 'parsed response:', [r for r in results]


qp.go
package main

import (
	"database/sql"
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"github.com/arnehormann/sqlinternals/mysqlinternals"
	_ "github.com/go-sql-driver/mysql"
	"log"
	"math/rand"
	"net/http"
	"os"
	"os/signal"
	"reflect"
	"regexp"
	"sync"
	"syscall"
	"time"
	_ "net/http/pprof"
	"runtime/debug"
)

// globals
// constants
const DECIMAL_AS_STRING = true
// configuration
var version string // NOTE: supplied automatically via Makefile
var debugLog bool
var maxDsns int
var maxConnsPerDsn int
var heapDumpDir string
// data
var dbs = make(map[string]*sql.DB)
var dbsMutex = &sync.Mutex{}

func dbFromDsn(dsn string) (db *sql.DB, err error) {
	// If the connections for this DSN are cached, return them
	dbsMutex.Lock()
	defer dbsMutex.Unlock()
	dbi, ok := dbs[dsn]
	if ok {
		return dbi, nil
	}
	if len(dbs)+1 > maxDsns {
		return nil, errors.New("dsn cache full")
	}
	dbc, err := sql.Open("mysql", dsn)
	if err != nil {
		log.Printf("Error initializing database connection: %s\n", err.Error())
		return nil, err
	}
	dbc.SetMaxIdleConns(maxConnsPerDsn)
	dbc.SetMaxOpenConns(maxConnsPerDsn)
	dbs[dsn] = dbc
	return dbc, nil
}

func flattenNullable (writeCols []interface{}) {
	// change nullable structs into values or nil
	for i, intf := range (writeCols) {
		kind := reflect.TypeOf(intf).Elem().Kind()

		// if the value scanned into a struct, it was nullable
		if (kind != reflect.Struct) {
			continue
		}

		valid := reflect.ValueOf(intf).Elem().FieldByName("Valid").Bool()
		if (valid) {
			// if it was nullable and not null, we just want the value
			nintf := reflect.ValueOf(intf).Elem().FieldByIndex([]int{0}).Interface()
			writeCols[i] = nintf
		} else {
			// if it was nullable and null, we want nil
			writeCols[i] = nil
		}
	}
}

type dsnResults struct {
	dsn string
	header []interface{}
	types []interface{}
	records [][]interface{}
}

func runQuery(dsn string, query string, results chan<- dsnResults) {
	var res dsnResults
	res.dsn = dsn
	defer func() {results <- res}()

	// get the specified database handle
	db, err := dbFromDsn(dsn)
	if err != nil {
		log.Printf("Error finding database (%s): %s\n", cleanNameFromDsn(dsn), err.Error())
		return
	}

	// execte the query
	if debugLog {
		log.Printf("Querying %s: %s\n", cleanNameFromDsn(dsn), query)
	}
	rows, err := db.Query(query)
	if err != nil {
		log.Printf("Error querying database: %s\n", err.Error())
		return
	}

	// get MySQL column information for interface types
	type DecimalString string;
	DecimalStringType := reflect.TypeOf(DecimalString(""));
	StringType := reflect.TypeOf(string(""));

	columns, err := mysqlinternals.Columns(rows)
	if err != nil {
		log.Printf("Error inspecting column types: %s\n", err.Error())
		return
	}
	sqlTypes := make([]reflect.Type, len(columns))
	for i, _ := range sqlTypes {
		res.header = append(res.header, columns[i].Name())
		sqlType, err := columns[i].ReflectSqlType(true)
		if err != nil {
			if DECIMAL_AS_STRING && columns[i].IsDecimal() {
				sqlType = DecimalStringType
			} else {
				log.Printf("Error creating storage for column type: %s\n", err.Error())
				return
			}
		}
		sqlTypes[i] = sqlType
		res.types = append(res.types, sqlType.Name())
	}

	// store the result set
	readCols := make([]interface{}, len(columns))
	writeCols := make([]interface{}, len(columns))
	for rows.Next() {
		for i, sqlType := range sqlTypes {
			if sqlType == DecimalStringType {
				/* This allows us to scan into a string,
				   but leave "DecimalString" in the type list */
				sqlType = StringType
			}
			writeCols[i] = reflect.New(sqlType).Interface()
			readCols[i] = writeCols[i]
		}
		err := rows.Scan(readCols...)
		if err != nil {
			log.Printf("Error scanning result record: %s\n", err.Error())
			results <- res
			return
		}

		flattenNullable(writeCols)

		outCols := make([]interface{}, len(columns))
		copy(outCols, writeCols)
		res.records = append(res.records, outCols)
	}

	// return the results
	return
}

func handleRequest(w http.ResponseWriter, r *http.Request) {
	resultsChan := make(chan dsnResults)

	// parse queries
	type DsnQuery struct {
		Dsn   string
		Query string
	}
	type Request struct {
		Flat bool
		Queries []DsnQuery
	}
	var req Request

	decoder := json.NewDecoder(r.Body)
	err := decoder.Decode(&req)
	if err != nil {
		log.Printf("Error decoding json: %s\n", err.Error())
		w.WriteHeader(500)
		return
	}

	queries := req.Queries
	if len(queries) == 0 {
		log.Printf("No queries requested\n")
		w.WriteHeader(500)
		return
	}

	// execute queries in parallel
	for i := range queries {
		go runQuery(queries[i].Dsn, queries[i].Query, resultsChan)
	}

	// aggregate responses
	var allResults []dsnResults
	for _ = range queries {
		allResults = append(allResults, <-resultsChan)
	}

	type flatResults [][]interface{}
	type mappedResults map[string][][]interface{}

	// map results to dsn names
	var results interface{}
	if req.Flat {
		results = make(flatResults, 2)
		results.(flatResults)[0] = allResults[0].header
		results.(flatResults)[1] = allResults[0].types
		for _, dsnRes := range allResults {
			results = append(results.(flatResults), dsnRes.records...)
		}
	} else {
		results = make(mappedResults)
		for _, dsnRes := range allResults {
			results.(mappedResults)[dsnRes.dsn] = dsnRes.records
		}
	}

	// send responses
	res, _ := json.Marshal(results)

	// send responses
	log.Printf("Sending %d responses to %d queries\n", len(allResults), len(queries))
	if debugLog {
		log.Printf("Response: %s\n", res)
	}
	w.Header().Set("Content-Type", "application/json")
	w.Write(res)
}

func cleanNameFromDsn(dsn string) (cleanName string) {
	re := regexp.MustCompile(".+@(.+)/(.+)\\??.*")
	nameParts := re.FindStringSubmatch(dsn)
	if nameParts != nil {
		cleanName = fmt.Sprintf("%s/%s", nameParts[1], nameParts[2])
	} else {
		cleanName = "<Couldn't parse host/database>"
	}
	return cleanName
}

func closeAll() {
	dbsMutex.Lock()
	defer dbsMutex.Unlock()
	for dsn, db := range dbs {
		log.Println("Closing:", dsn)
		db.Close()
	}
}

func signalHandler(s chan os.Signal) {
	for {
		sig := <-s
		log.Println("We get signal:", sig)
		if sig.String() == "user defined signal 1" {
			// dump heap to a file when USR1 is received
			filename := fmt.Sprintf("qp-heapdump-%d", time.Now().Unix())
			filepath := heapDumpDir + filename
			log.Printf("dumping heap to %s\n", filepath)
			f, err := os.Create(filepath)
			if err != nil {
				log.Printf("Couldn't open file for writing\n")
			} else {
				debug.WriteHeapDump(f.Fd())
			}
		} else {
			closeAll()
			os.Exit(1)
		}
	}
}

func main() {
	var versionFlag = flag.Bool("version", false, "print version")
	var dsnsFlag = flag.Int("maxDsns", 16, "maximum number of cached dsns")
	var connsFlag = flag.Int("maxConnsPerDsn", 24, "maximum number of connections per dsn")
	var socketFlag = flag.String("url", ":9666", "the socket url to listen to")
	var debugFlag = flag.Bool("debug", false, "log debug information")
	var heapDumpDirFlag = flag.String("heapDumpDir", "/mnt/tmp/", "directory in which to dump heap")
	flag.Parse()
	if *versionFlag {
		fmt.Println(version)
		os.Exit(0)
	}
	c := make(chan os.Signal, 1)
	signal.Notify(c, os.Interrupt, os.Kill, syscall.SIGUSR1)
	go signalHandler(c)

	socketUrl := *socketFlag
	maxDsns = *dsnsFlag
	maxConnsPerDsn = *connsFlag
	debugLog = *debugFlag
	heapDumpDir = *heapDumpDirFlag
	rand.Seed(time.Now().UnixNano())

	log.Printf("version:%s socket:%s maxDsns:%d debug:%t heapDumpDir:%s\n",
	           version, socketUrl, maxDsns, debugLog, heapDumpDir)

	http.HandleFunc("/", handleRequest)
	err := http.ListenAndServe(socketUrl, nil)
	if err != nil {
		log.Fatal("Can't serve: ", err)
	}
	closeAll()
	os.Exit(0)
}
